### ISSUE TITLE
boost/1.85.0 compile issues with msm::front::state_machine_def on llvm/clang++ 18.1.5 with std=c++20

### ISSUE THREAD

**Issue by hoyhoy** [boostorg/msm#72]
We have some code that uses

```
struct front_msm : msm::front::state_machine_def<front_msm, visitable_state> {...}
struct visitable_state {}
typedef boost::msm::back::recursive_get_transition_table<front_msm>::type recursive_stt;

typedef boost::msm::back::recursive_get_transition_table<Sm_rund_front>::type recursive_stt;
typedef boost::msm::back::generate_state_set<recursive_stt>::type all_states;
```

This is erroring out with boost 1.85.0 on the latest clang.  boost 1.84.0 compiles successfully though.

clang++ 18.1.5 errors out with this on the compile, but boost 1.84.0 compiles successfully.

```
/include/boost/msm/back/args.hpp:60:45: error: 'no_args' does not refer to a value [Semantic Issue]
   60 | BOOST_PP_REPEAT(BOOST_MSM_VISITOR_ARG_SIZE, MSM_VISITOR_ARGS, ~)
      |                                             ^
/include/boost/preprocessor/repetition/repeat.hpp:30:26: note: expanded from macro 'BOOST_PP_REPEAT'
   30 | # define BOOST_PP_REPEAT BOOST_PP_CAT(BOOST_PP_REPEAT_, BOOST_PP_AUTO_REC(BOOST_PP_REPEAT_P, 4))
      |                          ^
/include/boost/preprocessor/cat.hpp:22:32: note: expanded from macro 'BOOST_PP_CAT'
   22 | #    define BOOST_PP_CAT(a, b) BOOST_PP_CAT_I(a, b)
      |                                ^
/include/boost/preprocessor/cat.hpp:29:34: note: expanded from macro 'BOOST_PP_CAT_I'
   29 | #    define BOOST_PP_CAT_I(a, b) a ## b
      |                                  ^
<scratch space>:49:1: note: expanded from here
   49 | BOOST_PP_REPEAT_1
      | ^
/include/boost/msm/back/args.hpp:31:8: note: declared here
   31 | struct no_args {};
      |        ^
/include/boost/msm/back/args.hpp:60:45: error: use of undeclared identifier 'MSM_ARGS_PRINT' [Semantic Issue]
   60 | BOOST_PP_REPEAT(BOOST_MSM_VISITOR_ARG_SIZE, MSM_VISITOR_ARGS, ~)
      |                                             ^
/include/boost/preprocessor/repetition/repeat.hpp:30:26: note: expanded from macro 'BOOST_PP_REPEAT'
   30 | # define BOOST_PP_REPEAT BOOST_PP_CAT(BOOST_PP_REPEAT_, BOOST_PP_AUTO_REC(BOOST_PP_REPEAT_P, 4))
      |                          ^
/include/boost/preprocessor/cat.hpp:22:32: note: expanded from macro 'BOOST_PP_CAT'
   22 | #    define BOOST_PP_CAT(a, b) BOOST_PP_CAT_I(a, b)
      |                                ^
/include/boost/preprocessor/cat.hpp:29:34: note: expanded from macro 'BOOST_PP_CAT_I'
   29 | #    define BOOST_PP_CAT_I(a, b) a ## b
      |                                  ^
<scratch space>:49:1: note: expanded from here
   49 | BOOST_PP_REPEAT_1
      | ^
```

**Comment by hoyhoy** [comment]
Seems like `args.hpp` is missing `#include <boost/preprocessor/repetition/enum.hpp>`

**Comment by hoyhoy** [comment]
@ChristophStrehle there are breaking changes in boost/1.85.0.  MSM now processes deferred events and the event table differently than boost/1.84.0.  Both clang and gcc segmentation fault with our codebase written in 2017.

We have an event that is changing state and adding a deferred event and that caused an infinite loop.  I just refactored our code to prevent the state change by deferring the second event rather than moving to the next state.  The behavior of the msm changed between 1.84.0 and 1.85.0 for us.

**Comment by ChristophStrehle** [comment]
Hello @hoyhoy, unfortunately I'm not so familiar with the implementation of the msm state machine. From the callstack and the description you provided I would guess that the change in behavior is caused by the commit f37d4a88f1dfd9821859d2f3f2f91b3b2b6a3cc8 from @henry-ch.

**Comment by henry-ch** [comment]
Hello @hoyhoy,
could you provide me with a minimum example? It is possible that I induced a new bug while fixing one, it might also be possible that the fsm was wrong.

**Comment by hoyhoy** [comment]
Yeah, it's mega broken for us right now.   Whatever was done post 1.83 seemed to break for one of our applications.

https://developercommunity.visualstudio.com/t/cl-193933523-x64-boost1840-boost/10622335

We can't compile on MSVC, and the library doesn't change states in the way that it used to.  I have a workaround, but we're actually looking at removing it entirely since Boost doesn't appear to unit test before releasing.  Is MSM abandoned?

I posted a minimum repro above.  Basically, if you defer event, Boost 1.85 changed how the state machine worked.  I'm imagining you have no tests for this.  But you just inadvertently broke the MSM.  The bug is, `handle_defer_helper()` is not changing the next state when it didn't do that previously.

**Comment by henry-ch** [comment]
I cannot understand the problem with the above code. Sorry, "mega broken" is not helpful. Please provide a complete example with the whole state machine. Guessing is usually not a good idea.
I cannot reproduce the compiler crash with VS 17.8 (and Microsoft also not from what I see). Which compiler are you using?
Please provide also there a complete example, if possible, I will try to find the compiler you are using.

**Comment by hoyhoy** [comment]
> The bug does still happen on the latest commit, tested with all three backends.

We had to refactor our msm implementation to defer an event instead of process it to avoid the infinite loop.  

The bug is that `handle_defer_helper` is changing the to the next state when it didn't do that in boost/1.84.0.  This is causing an infinite loop.  Still not fixed over a year later.

**Comment by chandryan** [comment]
> Allow me to reflect on your behavior -- I stated _exactly_ what the bug in msm was over a year ago i.e. `handle_defer_event()` is changing to the next state in 1.85, whereas in 1.84 it didn’t.

While you are very welcome to share suggestions about where the root cause might be, you were explicitly requested to share a complete example for reproducing the issue. Looking through the conversation within this issue, your missing response seems to be the main reason this bug has not been worked on for such a long time.

And if you happen to know _exactly_ what the bug is, you are always invited to create a PR containing the bugfix and a regression test to contribute to the library.

> I’m not expecting anything other than for msm to _not_ randomly merge breaking changes without a deprecation notice.

You _can_ expect library developers to _not_ do something like this intentionally.
You _cannot_ expect them to not accidentally break something.
 
> [...] because I assumed (rightly) that boost::msm doesn't have unit test coverage around this behavior. So, nobody would know it was broken until it was released.

You _can_ expect library developers to follow Best Practices and add regression tests for identified & resolved bugs. 
You _cannot_ expect an OSS library, or any piece of software in this regard, to have unit tests covering all possible cases.

---

I commented about your statement in this issue in order to motivate you to reflect about how your behavior can be perceived by others, especially the ones you are asking for support.

You complaining to developers about an *accidentally* broken functionality in an OSS lib, which you can use free-of-charge, brings nothing but frustration to all involved parties. What do you get out of this? Best case you were able to get off some steam at the cost of agitating others, worst case you demotivate the people you are asking for support to look into your problem.

I'm asking you no more than to consider how your behavior affects others, because nonviolent & respectful communication is essential - especially in community-driven OSS projects. And statements such as "mega broken", "still not fixed over a year later" or "randomly merge breaking changes" are simply inappropriate and will not help anyone.

Take it or leave it, I will not comment further on your behavior.

**Comment by hoyhoy** [comment]
You're right, I should have provided an example, but it would have taken me hours to do that.  Our msm code is buried in a  closed-source app.

I just refactored out our reliance on this  `handle_defer_helper()` behavior.  This was one problem of about 5,000 problems we faced when updating our dependencies last year.

**Comment by chandryan** [comment]
I looked into the background of the implementation change, it was related to #18. As far as I understand, it attempts to retain the correct order of deferred events in the queue. The algorithm goes through all deferred events in the queue and sorts it, whenever an event was successfully processed. The recently appended items go back to the front of the queue and then the algorithm starts from the beginning.

This algorithm goes into infinite recursion because of the orthogonal regions in your example: The event is processed and deferred at the same time, which means it goes back into the queue but also sorted, and then processed again and again. 

I have [one idea in mind](https://github.com/chandryan/msm/pull/14/files) how to remove the infinite recursion and probably even retain the correct event order in *some* cases, but this feels more like a makeshift solution. That idea relies on the assumption, that at max. 1 region is deferring the event. If that algorithm were applied in a situation where both regions defer the event, the sorting would not work anymore. 

In general I cannot think of a fully working implementation for event deferral, if orthogonal regions are involved and the event has to be dispatched for evaluation. I don't know what your use case is @shingst , but there is a high chance that even without infinite recursion the deferral might not do what you want:
In your example Event1 gets consumed and deferred at the same time. As soon as State2 is not active anymore, State4 would consume the same Event1 a second time.
Other side effects would come into place, if you had a situation where both regions defer the event; its occurrences in the deferred events queue would increase with every attempt to process them.

I would like to look into a UML-compliant solution for the new back-end `backmp11`, which is described in #143. Unfortunately I don't see a way to implement the same in the older back-ends, the reasons for this are also described in #143.

I'm not sure how to go forward with `back` and `back11`, I see the following possible options:
- Go with the makeshift solution to prevent infinite recursion and have the correct event order in *some* cases,
- Revert the implementation to the previous state without the attempt to sort,
- Or find some other solution that I cannot see.

What's your opinion @henry-ch ?

**Comment by henry-ch** [comment]
Please don't! Conditional event deferring is one of the feature which was popular when MSM came out. I personally use it quite often. And in that case, one will need to defer as an action, which is fine because there is a guard.
I think removing widely used features because one can use it wrong goes against the idea of C++ which, well, will let you blow your leg if you want to ;-)
